### DISCUSSION TITLE
When using multiple remote certificates only a single identity is supported

### DISCUSSION THREAD

**Discussion by johnfzc** [strongswan/strongswan#2029] [Q&A] [answered]
I am replacing a VPN hub router and the subject of the certificate on the router will change.

The spokes connected to this hub are running a number of different versions of strongswan: 5.9.0, 5.9.2 and 5.9.5.

In my current swanctl configuration on the spokes we use a remote identity identified using asn1. We cannot use SANS and the way that Cisco formulates the asn1 structure of the subject of the device requires us to use this notation.

```
connections {
   tun2 {
      [...]
      remote {
         auth = pubkey
         id = "asn1dn:#massive-hex-string"
      }
```

I would like to support both the old and new certificates while transitioning the hardware.

I found a [mailing list discussion](https://lists.strongswan.org/pipermail/users/2018-May/012631.html) where people spoke about creating an id1 to address this type of problem.

When I tried this I got the following error:

```
loading connection 'tun2' failed: unknown option: id1, config discarded
```

I also see that the `certs` configuration should allow us to specify multiple permitted certificates for a remote peer. After loading the old and new certificate into the `/etc/swanctl/x509/` directory I configured two certificates in a comma separated list as follows `certs = "hub2-old.pem,hub2-new.pem"`. This is correctly parsed, and when I list the configuration using `--list-conns` I see the two certificates present. I also see that a single id is present; it appears to be the subject of the first certificate provided in the list; when I change the order of the certificates in the list the id changes as well.

```
root@spoke:/tmp# swanctl --list-conns
[...]
tun2: IKEv2, no reauthentication, rekeying every 14400s, dpd delay 920s
[...]
  remote public key authentication:
    id: CN=router2.example.com, unstructuredName=router2.example.com
    certs: CN=router2.example.com, unstructuredName=router2.example.com CN=router2.example.com, serialNumber=FAILFAILFAIL, unstructuredName=router2
```

If I then try to initiate the connection with the second certificate from the list authentication fails as follows:

```
root@spoke:/tmp# ipsec up tun1464
[...]
authentication of 'CN=router2.example.com, serialNumber=FAILFAILFAIL, unstructuredName=router2' with RSA signature successful
constraint check failed: identity 'CN=router2.example.com, unstructuredName=router2.example.com' required
selected peer config 'tun2' unacceptable: constraint checking failed
```

How can I allow my spokes to support two different certificates and identities for their right side peer?

**Comment by tobiasbrunner** [comment, marked as answer]
> I found a [mailing list discussion](https://lists.strongswan.org/pipermail/users/2018-May/012631.html) where people spoke about creating an id1 to address this type of problem.

That's for secrets, not for identities in remote sections.

> I also see that a single id is present; it appears to be the subject of the first certificate provided in the list; when I change the order of the certificates in the list the id changes as well.

You can set `remote.id` to `%any` to avoid the fallback to the subject DN of the first certificate.

**Reply by johnfzc** [reply]
Awesome, I think that will work, I'm going to test it immediately and report back. Update: After testing I can confirm that this works!

Can you please confirm that, because the certs are specified the devices will trust any identity as long as it presents one of the trusted certificates. It will not trust absolutely any certificate/identity.

Thank you very much Tobias, greatly appreciated.

**Reply by johnfzc** [reply]
I performed another test where I only included the certificate for the new peer and then attempted to connect to the old hub. strongSwan refused to establish a connection as I expected. Specifying `id = "%any"` in the remote section of my tunnel configuration resolved the problem I was having and only permits connections with the expected certificates specified in the remote configuration.

Thanks again for your help!

**Reply by tobiasbrunner** [reply]
Yes, as you noticed, if you explicitly configure certificates, the peers have to use one of them to get authenticated. The identity is irrelevant in this case as it can't be anything that's not contained in the used certificate, so configuring `%any` is perfectly fine.
